8277817: java/awt/dnd/BadSerializationTest/BadSerializationTest.java failed: ClassNotFoundException: com.apple.laf.AquaImageFactory$SystemColorProxy

cushon opened the pull request:
This change updates the serialized objects used by `java/awt/dnd/BadSerializationTest/BadSerializationTest.java` using a similar approach to the previous fix in [JDK-8039082](https://bugs.openjdk.java.net/browse/JDK-8039082).

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change must be properly reviewed
- [ ] Change requires a CSR request to be approved

### Issues
 * [JDK-8277817](https://bugs.openjdk.java.net/browse/JDK-8277817): java/awt/dnd/BadSerializationTest/BadSerializationTest.java failed: ClassNotFoundException: com.apple.laf.AquaImageFactory$SystemColorProxy
 * [JDK-8278555](https://bugs.openjdk.java.net/browse/JDK-8278555): java/awt/dnd/BadSerializationTest/BadSerializationTest.java failed: ClassNotFoundException: com.apple.laf.AquaImageFactory$SystemColorProxy (**CSR**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6603/head:pull/6603` \
`$ git checkout pull/6603`

Update a local copy of the PR: \
`$ git checkout pull/6603` \
`$ git pull https://git.openjdk.java.net/jdk pull/6603/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6603`

View PR using the GUI difftool: \
`$ git pr show -t 6603`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6603.diff">https://git.openjdk.java.net/jdk/pull/6603.diff</a>

</details>

prrace commented:
I must be losing it since I can't find a code review for the fix for https://bugs.openjdk.java.net/browse/JDK-8039082
Presumably the "fix" is to regenerate the serialized object files on some other platform beside mac ?
But 8039082 was supposed to have done that, yet we clearly have a macOS class in here.
I question the whole wisdom of including any Swing related classes in such a serialized form since Swing only supports short-term serialization. Not across platforms, not across releases. But still .. 

And isn't it just one of these cases that failed ? Did you have to re-generate all of them ?

On which platform did you re-generate these ? And on which platforms did you then test the result ?
I don't want to replace one failure with another.

prrace commented:
Also the test is now being problem-listed. https://github.com/openjdk/jdk/pull/6624
So this fix will need to de-problem list it.

cushon commented:
I couldn't find the review either.

I regenerated it by following the instructions to compile and run the test class as a standalone tool on `x86_64 GNU/Linux`, which updated all of the existing files. I verified it fixes the test on that platform, I don't have a great way to test it on the other platforms right now.

* https://github.com/openjdk/jdk/blob/5a4a9bb9d55134deac0e02cf37f31d1dd2223024/test/jdk/java/awt/dnd/BadSerializationTest/BadSerializationTest.java#L64-L65

* https://github.com/openjdk/jdk/blob/5a4a9bb9d55134deac0e02cf37f31d1dd2223024/test/jdk/java/awt/dnd/BadSerializationTest/BadSerializationTest.java#L95-L96

I started investigating this when I thought it was related to [JDK-8271623](https://bugs.openjdk.java.net/browse/JDK-8271623), if anyone wants to take this over (especially now that it's problem-listed) please feel free :)

mrserb commented:
As of the usefulness of the serializations tests in Swing - it is a good coverage of the cleaning code in the L&fs. Each L&F should clean all its data when the app switch one LF to another, but if old L&F forget to cleanup then the app will get a memory leak.

mrserb commented:
> I regenerated it by following the instructions to compile and run the test class as a standalone tool on `x86_64 GNU/Linux`, which updated all of the existing files. I verified it fixes the test on that platform, I don't have a great way to test it on the other platforms right now.

As far as I remember the previous fix regenerated the binary files only, for a specific reason - before that a couple of commits fixed a memory leak which prevents the Aqua L&F deletes all its delegates during cleanup step. This cleanup step is executed in two cases, when we serialize the component or when we switch one L&F to another.

Since the fix for JDK-8276665 exposed another bug you will need:
 1. Regenerate the data on macOS
 2. If the test will fail while use the new regenerated data, then you will need to fix the Aqua to properly delete the references to the Aqua specific objects.

cushon commented:
Thanks @mrserb! I regenerated the binary files on macOS, and that caused the test to fail on linux.

I don't have a lot of context here, but as far as I can tell `BasicPanelUI` was installing (but not uninstalling) a reference to the color `Panel.background`, which for Aqua is an instance of `com.apple.laf.AquaImageFactory$SystemColorProxy`.

* https://github.com/openjdk/jdk/blob/a363b7b9217cbb9a7580a87b812da8d5a4215326/src/java.desktop/share/classes/javax/swing/plaf/basic/BasicPanelUI.java#L81-L83

* https://github.com/openjdk/jdk/blob/4efcd2006f00a6347987ecdacd6e09d59a0006f4/src/java.desktop/macosx/classes/com/apple/laf/AquaLookAndFeel.java#L391-L392

* https://github.com/openjdk/jdk/blob/4efcd2006f00a6347987ecdacd6e09d59a0006f4/src/java.desktop/macosx/classes/com/apple/laf/AquaImageFactory.java#L469-L471

I included a fix to `BasicPanelUI`, but the test now fails with a different issue:

```
Caused by: java.lang.ClassNotFoundException: com.apple.laf.AquaFonts$DerivedUIResourceFont
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
```

prrace commented:
> As of the usefulness of the serializations tests in Swing - it is a good coverage of the cleaning code in the L&fs. Each L&F should clean all its data when the app switch one LF to another, but if old L&F forget to cleanup then the app will get a memory leak.

This test doesn't switch L&Fs .. if it wasn't for mac defaulting to Aqua they'd all be using Metal.

So this is an accidental by-product, not the goal of the test.

And that goal be met by the test writing out the serialized objects and reading them back all at "test time", can't it ?

Seems to me right now we are stumbling over cases where what is serialized is inappropriate for persistent serialization.

cushon commented:
Ok the other issue was another instance of the same problem with `BasicPanelUI`.

I now have the test passing on both linux and mac, after regenerating the binary files on mac.

I'm still not sure if this is the best approach, does this fix belong in `BasicPanelUI`? You mentioned fixing 'Aqua to properly delete the references to the Aqua specific objects', does it make sense to handle this in aqua specific code instead?

mrserb commented:
> This test doesn't switch L&Fs .. if it wasn't for mac defaulting to Aqua they'd all be using Metal.

It does not change the L&F but it will trigger the same cleanup step.

> So this is an accidental by-product, not the goal of the test.

yes we found one additional product bug.

> And that goal be met by the test writing out the serialized objects and reading them back all at "test time", can't it ?

No it cannot write and read it back in during test execution, the binary files should be created on the "patched" version of JDK.

mrserb commented:
> Ok the other issue was another instance of the same problem with `BasicPanelUI`.
> 
> I now have the test passing on both linux and mac, after regenerating the binary files on mac.

> I'm still not sure if this is the best approach, does this fix belong in `BasicPanelUI`? You mentioned fixing 'Aqua to properly delete the references to the Aqua specific objects', does it make sense to handle this in aqua specific code instead?

Usually the delegate which install some properties/listeners/etc is responsible for cleanup.

cushon commented:
> Usually the delegate which install some properties/listeners/etc is responsible for cleanup.

I'm not sure I understand, aren't these being installed in `BasicPanelUI`?

https://github.com/openjdk/jdk/blob/4166dfd004d16769f8ea1d3441d3d19dedfd6867/src/java.desktop/share/classes/javax/swing/plaf/basic/BasicPanelUI.java#L81-L85

Would it make sense to create a corresponding `LookAndFeel.installColorsAndFont` method that would be called from `BasicPanelUI. uninstallDefaults`?

(Sorry for the naive questions here--as I mentioned I started investigating this when I thought it was related to JDK-8271623, and am not familiar with this code.)

mrserb commented:
> I'm not sure I understand, aren't these being installed in `BasicPanelUI`?

If it is installs in the BasicListUI only then yes the fix belongs to this class. These properties changing also in some other classes as well, but some other classes cleanup it properly for example see BasicListUI.uninstallDefaults(), but I am not sure that all of them do this.

cushon commented:
> These properties changing also in some other classes as well, but some other classes cleanup it properly for example see BasicListUI.uninstallDefaults(), but I am not sure that all of them do this.

I looked through other uses of `installColorsAndFont`, and most of them don't seem to be uninstalling the properties. I updated the ones I found, and added a `uninstallColorsAndFont` method to avoid copying the logic around.

mrserb commented:
I think it is fine, need to run the tests for the change since it can uncover some hidden issues.

kevinrushforth commented:
@cushon As a best practice, please do not force push to a branch for which there is an PR under review. Instead, you can merge the master branch into your branch if you need to update.

cushon commented:
/csr

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/LookAndFeel.java:
```diff
@@ -187,6 +187,25 @@ public static void installColors(JComponent c,
+     * @see #uninstallColorsAndFont
+     * @throws NullPointerException if {@code c} is {@code null}
+     */
+    public static void uninstallColors(JComponent c) {
```
It's a new method, `@since 19` should be added.

aivanov-jdk commented on src/java.desktop/share/classes/javax/swing/LookAndFeel.java:
```diff
@@ -217,6 +236,22 @@ public static void installColorsAndFont(JComponent c,
+     * @param c component to uninstall the properties from
+     * @throws NullPointerException if {@code c} is {@code null}
+     */
+    public static void uninstallColorsAndFont(JComponent c) {
```
Should also have `@since 19`.

cushon replied:
Done

cushon commented:
I started a thread on client-libs-dev and filed a CSR: https://bugs.openjdk.java.net/browse/JDK-8278555

aivanov-jdk reviewed (commented):
@mrserb mentioned `BasicListUI.uninstallDefaults` uninstalls the colors correctly. However, with the new methods added, it could use `LookAndFeel.uninstallColorsAndFont` to reset its font, foreground and background.

Probably other UI classes should be updated too.

cushon commented:
@aivanov-jdk thanks, I updated `BasicListUI`.

I found a few other cases with  the following, which I have fixed:

```
$ find src/java.desktop/share/ -type f -name '*.java' | xargs grep -l LookAndFeel.installColors | xargs grep -L LookAndFeel.uninstallColors
```

The remaining files that call `LookAndFeel.installColors*` with a corresponding `uninstall*` call are below.

This reference in javadoc seems fine:
 https://github.com/openjdk/jdk/blob/8eb453baebe377697286f7eb32280ca9f1fd7775/src/java.desktop/share/classes/javax/swing/plaf/synth/SynthSpinnerUI.java#L116
 
This may be suspicious, but it isn't in an `install` method:

https://github.com/openjdk/jdk/blob/8eb453baebe377697286f7eb32280ca9f1fd7775/src/java.desktop/share/classes/sun/swing/FilePane.java#L1295

aivanov-jdk commented:
I've been contemplating about what is being done in this PR… It feels wrong.

> > As of the usefulness of the serializations tests in Swing - it is a good coverage of the cleaning code in the L&fs. Each L&F should clean all its data when the app switch one LF to another, but if old L&F forget to cleanup then the app will get a memory leak.

Swing explicitly warns, “Serialized objects of this class will not be compatible with future Swing releases.”

Don't break this rule in this test? A component is serialized using one version of the JDK and is deserialized in another version of the JDK and even on another platform.

> This test doesn't switch L&Fs .. if it wasn't for mac defaulting to Aqua they'd all be using Metal.
> 
> So _this is an accidental by-product_, not the goal of the test.

It looks so.

If the L&Fs are expected to clean up when uninstalled, it should be tested thoroughly for all the components and Look-and-Feels. Only this way we can be sure the properties get reset.

While working on this code review, Liam has found and fixed quite a few places where clean-up wasn't done properly.

> And that goal be met by the test writing out the serialized objects and reading them back all at "test time", can't it ?
> 
> Seems to me right now we are stumbling over cases where what is serialized is inappropriate for persistent serialization.

Taking into account the above comments, this test could be simplified to write out the objects and then read them in. No persistent serialization which is _explicitly_ unsupported.


To be clear, I think the proposed change is good on its own: it adds convenience methods to clean up UI properties when a L&F is uninstalled, it adds the clean-up where it was missing. But it doesn't fit into the fix of this test failure. This work should be an enhancement performed under another bugid.

prrace commented:
Whilst there might well be some need to do something to make sure uninstallUI() is doing its job entirely/properly,
I am not sure that I'd want to push such a change as the resolution to this failing test.

Better to start afresh with a new bug with an appropriate symopsis and then write some new tests that verify the UI is uninstalled on
all the Swing components as you cycle though available L&Fs per-platform.

And since the problem clearly exists in older versions of JDK (even if not exposed by the recent change in JDK 18),
then how would we resolve it there since clearly the new API can't be backported.

And perhaps new API isn't actually needed ? Are there other options ?

And as I wrote before, the failing test is not 100% valid since Swing's specification says :

"Warning: Serialized objects of this class will not be compatible with future Swing releases.
The current serialization support is appropriate for short term storage or RMI between applications running the same version of Swing."

"Same version" could be interpreted as "Swing in JDK N", or "Swing in JDK N on <specific platform>"
But either way, it isn't valid to check in a serialized form from Swing version N and have *any* expectation that it will run on N+1,
which is what this test does.
Yes, it can be argued that this test can/will fail even between platforms on the same version, but (1) I'm questioning whether
even that serialization was intended and (2) the test is not limiting itself to that scenario.

I did not understand the "special JDK" answer as to why this test needs to have a checked in serialized form on disk at all.
If we can't do that, I think we should scrap the test or find a different way to do whatever it intends to verify.
We should not be treating it as a test of uninstalling UI - it is a mere accident that this cropped up.

cushon commented:
> > To be clear, I think the proposed change is good on its own: it adds convenience methods to clean up UI properties when a L&F is uninstalled, it adds the clean-up where it was missing. But it doesn't fit into the fix of this test failure. This work should be an enhancement performed under another bugid.
>
> Better to start afresh with a new bug with an appropriate synopsis and then write some new tests that verify the UI is uninstalled on all the Swing components as you cycle though available L&Fs per-platform.

I have filed https://bugs.openjdk.java.net/browse/JDK-8278620 to track adding test coverage for the `uninstallUI` methods and deciding on the best fix for that issue.

> Taking into account the above comments, this test could be simplified to write out the objects and then read them in. No persistent serialization which is explicitly unsupported.
>
>  I think we should scrap the test or find a different way to do whatever it intends to verify.

Currently the test relies on making a manual change to `DragGestureEvent` and `DragGestureRecognize` before serializing the objects, so it's not clear to me how the test would be modified to do that at test-time: https://github.com/openjdk/jdk/blob/5a4a9bb9d55134deac0e02cf37f31d1dd2223024/test/jdk/java/awt/dnd/BadSerializationTest/BadSerializationTest.java#L95-L96

I'm happy to update this change to scrap the test if there's agreement that's the best resolution.

mrserb commented:
> I'm happy to update this change to scrap the test if there's agreement that's the best resolution.

I disagree to do that. First, the test is quite important to check the broken serialization form, see the initial bug report. Second, the test proved to be useful since it found two additional bugs.

cushon commented:
Ok!

So is the idea to wait for a separate fix for [JDK-8278620](https://bugs.openjdk.java.net/browse/JDK-8278620), and then update the checked-in serialized objects and re-enable the test?

What do you think of the suggestion to simplify the test to write out the objects and then read them in, without persistent serialization, since there was some discussion about that being explicitly unsupported? I'm still not sure how that would be achieved with the current test, since it relies on modifications to `DragGestureEvent` and `DragGestureRecognize` to serialize the objects?

aivanov-jdk commented:
I see the problem: the constructors do not allow setting invalid values to the fields. This is what the *special version of JDK* refers to, the build where the checks for invalid values are removed from the constructors. It explains why the serialized form can't be obtained by writing an object out, it's impossible to create such an object.

I wonder if it's possible to modify the value of the field using reflection via `setAccessible(true)`. With strong encapsulation, it may not work. If it works, the objects with invalid can be serialized on the fly.

cushon commented:
I have modified the test to create the serialized objects at test-time, instead of relying on checking in the serialized files.

> I wonder if it's possible to modify the value of the field using reflection via `setAccessible(true)`

Done: the test now creates valid objects, and then uses reflection to set the fields to `null` after the constructor has executed.

cushon commented:
/csr unneeded

cushon commented:
> only Reviewers can determine that a CSR is not needed.

I marked this as requiring a CSR when the PR included a fix for https://bugs.openjdk.java.net/browse/JDK-8278620 that added new methods to `LookAndFeel`. This fix is now only changing the test, which should not require a CSR.

cushon commented:
> just regenerate the data in this test

Based on https://github.com/openjdk/jdk/pull/6603#issuecomment-991295599 and https://github.com/openjdk/jdk/pull/6603#issuecomment-991311158, I thought we didn't want to rely on checking in the serialized data.

mrserb commented:
I do not see a problem to have that serialized data since it have a good coverage and proved to be able find a bugs. Note that all time this test fails were unrelated to the versioning of Swing, so I do not see a reason why we should change that to the version which will not be able find that bugs.

I agree that it would be better to have a tests which will found such issue, but looks like we do not have such. For example for the current issue I think nobody could imaging that there are a cleaning issue of fonts/colors/etc.

mrserb commented:
I do not see a problem to have that serialized data since it have a good coverage and proved to be able find bugs. Note that every time this test fails were unrelated to the versioning of Swing, so I do not see a reason why we should change that to the version which will not be able find that bugs.

I agree that it would be better to have a tests which will found such issue, but looks like we do not have such. For example for the current issue I think nobody could imaging that there are a cleaning issue of fonts/colors/etc.

Since we found a real bugs, I suggest to thinking about an opposite solution and save the data not only on mac but on all platforms, and then load all of them during test execution, eventually this is a test for DnD in AWT which may also some issue related to peers and native objects.

cushon commented:
I don't have a strong preference between the current approach in the PR of not persisting serialized objects, and of updating the existing serialized objects; whichever the `client-libs` group prefers is fine with me.

I restored the earlier version of the PR that updated the existing serialized data as https://github.com/openjdk/jdk/pull/7100

> > I wonder if it's possible to modify the value of the field using reflection via `setAccessible(true)`
>
> Done: the test now creates valid objects, and then uses reflection to set the fields to `null` after the constructor has executed.

@aivanov-jdk I spun that cleanup out into a separate issue, since I think it's a worthwhile improvement regardless of which fix is accepted here: https://github.com/openjdk/jdk/pull/7099

cushon commented:
Hi, is there any feedback on this change, or the alternate fix in #7100 that updates the binary files, or the related cleanup in #7099?

mrserb commented:
I guess the discussion above suggests extracting the "big" change to the Swing package you did to the separate fix, I missed in what PR that change was moved?

cushon commented:
> Hi, is there any feedback on this change, or the alternate fix in #7100 that updates the binary files, or the related cleanup in #7099?

Friendly ping.

My opinion is that https://github.com/openjdk/jdk/pull/7099 is a clear improvement to the current test setup, because it preserves the existing behaviour without relying on patching the JDK to disable constructor validation in `DragGestureEvent` and `DragGestureRecognizer` to generate the serialized objects.

Additionally, I think it's worth considering the fix in https://github.com/openjdk/jdk/pull/7100. It includes a minimal fix to the `uninstallDefaults` behaviour of `BasicPanelUI`, which is the only broken uninstall behaviour exercised by the current test, and allows moving the test off the allowlist.

However if the decision is that those approaches aren't worth pursuing, and it would be better to wait to set up comprehensive test coverage for the uninstall bugs in this package ([JDK-8278620](https://bugs.openjdk.java.net/browse/JDK-8278620)), I'm happy to drop these PRs.

Thanks!

cushon commented:
I'm going to close this PR, since there doesn't seem to be a path forward here.
